File: module/dict_tools.py
from copy import deepcopy


def deep_merge(a: dict, b: dict) -> dict:
    
    result = deepcopy(a)

    for b_level, b_source_list in b.items(): #b_level=b-key; b_source_list=b-value  Levels

        for b_source in b_source_list:

            if b_level not in result.keys():
                result[b_level] = []

            if b_source['source'] in [item['source'] for item in result[b_level]]:
                continue

            result[b_level].append(b_source)

    return result



def deep_list_merge(a: [], b: []) -> []:
    """[
            {
                'level': 3, 
                'sources': 
                    [
                        {
                            'source': 'prouzalib/qrpglesrc/cpysrc2ifs.sqlrpgle.pgm', 
                            'cmds': []
                        }, 
                        {'source': ...
    """
    
    result = deepcopy(a)

    for b_level_item in b:

        b_level = b_level_item['level']
        b_sources = b_level_item['sources']

        result_level_item = [item for item in result if item['level'] == b_level]
        if result_level_item:
            result_level_item = result_level_item[0]

        if not result_level_item:
            result_level_item={'level': b_level, 'sources': []}

        for b_source in b_sources: #b_level=b-key; b_source_list=b-value  Levels

            if b_source['source'] in [item['source'] for item in result_level_item['sources']]:
                continue

            #result_level_item['sources'].append({'source': b_source['source'], 'cmds': b_source['cmds']})

            found = False
            for i in range(len(result)):
                if b_level != result[i]['level']:
                    continue
                found = True
                result[i]['sources'].append({'source': b_source['source'], 'cmds': b_source['cmds']})
                break

            if not found:
                result.append({'level': b_level, 'sources': [{'source': b_source['source'], 'cmds': b_source['cmds']}]})

    return result




def dict_merge(base_dct, merge_from_dct):
    for k in set(base_dct.keys()).union(merge_from_dct.keys()):
        if k in base_dct and k in merge_from_dct:
            if isinstance(base_dct[k], dict) and isinstance(merge_from_dct[k], dict):
                yield (k, dict(dict_merge(base_dct[k], merge_from_dct[k])))
            else:
                if isinstance(merge_from_dct[k], bool) or len(merge_from_dct[k]) != 0:
                    yield (k, merge_from_dct[k])
                else:
                    yield (k, base_dct[k])
        elif k in base_dct:
            yield (k, base_dct[k])
        else:
            yield (k, merge_from_dct[k])

File: src/module/dict_tools.ts
// TypeScript conversion of module/dict_tools.py

interface SourceItem {
    source: string;
    cmds: any[];
}

interface LevelItem {
    level: number;
    sources: SourceItem[];
}

export function deepMerge(a: Record<string, SourceItem[]>, b: Record<string, SourceItem[]>): Record<string, SourceItem[]> {
    const result = { ...a };

    for (const [bLevel, bSourceList] of Object.entries(b)) {
        for (const bSource of bSourceList) {
            if (!(bLevel in result)) {
                result[bLevel] = [];
            }

            const sourceExists = result[bLevel].some(item => item.source === bSource.source);
            if (!sourceExists) {
                result[bLevel].push(bSource);
            }
        }
    }

    return result;
}

function deepClone<T>(obj: T): T {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(item => deepClone(item)) as any;
    }
    const cloned: any = {};
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            cloned[key] = deepClone(obj[key]);
        }
    }
    return cloned;
}

export function deepListMerge(a: LevelItem[], b: LevelItem[]): LevelItem[] {
    const result = deepClone(a);

    for (const bLevelItem of b) {
        const bLevel = bLevelItem.level;
        const bSources = bLevelItem.sources;

        let resultLevelItem = result.find((item: LevelItem) => item.level === bLevel);

        if (!resultLevelItem) {
            resultLevelItem = { level: bLevel, sources: [] };
            result.push(resultLevelItem);
        }

        for (const bSource of bSources) {
            const sourceExists = resultLevelItem.sources.some((item: SourceItem) => item.source === bSource.source);
            if (!sourceExists) {
                resultLevelItem.sources.push({ source: bSource.source, cmds: bSource.cmds });
            }
        }
    }

    return result;
}

export function dictMerge(baseDct: any, mergeFromDct: any): any {
    const result: any = {};
    const allKeys = new Set([...Object.keys(baseDct), ...Object.keys(mergeFromDct)]);

    for (const k of allKeys) {
        if (k in baseDct && k in mergeFromDct) {
            if (typeof baseDct[k] === 'object' && baseDct[k] !== null && 
                typeof mergeFromDct[k] === 'object' && mergeFromDct[k] !== null &&
                !Array.isArray(baseDct[k]) && !Array.isArray(mergeFromDct[k])) {
                result[k] = dictMerge(baseDct[k], mergeFromDct[k]);
            } else {
                if (typeof mergeFromDct[k] === 'boolean' || 
                    (typeof mergeFromDct[k] === 'string' && mergeFromDct[k].length !== 0) ||
                    (typeof mergeFromDct[k] !== 'string' && mergeFromDct[k])) {
                    result[k] = mergeFromDct[k];
                } else {
                    result[k] = baseDct[k];
                }
            }
        } else if (k in baseDct) {
            result[k] = baseDct[k];
        } else {
            result[k] = mergeFromDct[k];
        }
    }

    return result;
}
